sql/columndefault.go

```diff
@@ -0,0 +1,141 @@
+package sql
+
+import (
+	"fmt"
+	"strings"
+)
+
+type columnDefaultManager struct {
+	AllowedDefaultFunctions FunctionRegistry
+	ExpressionTransformUp func(Expression) (Expression, error)
+	ParseDefaultString func(*Context, string) (Expression, error)
+}
+var ColumnDefaultManager = &columnDefaultManager{
+	AllowedDefaultFunctions: NewFunctionRegistry(),
+}
+
+// StringToColumnDefaultValue takes in a string representing a default value and returns the equivalent Expression.
+func StringToColumnDefaultValue(ctx *Context, exprStr string) (*ColumnDefaultValue, error) {
+	parsedExpr, err := ColumnDefaultManager.ParseDefaultString(ctx, exprStr)
+	if err != nil {
+		return nil, err
+	}
+	// The literal and expression distinction seems to be decided by the presence of parentheses, even for defaults like NOW() vs (NOW())
+	// 2+2 would evaluate to a literal under the parentheses check, but will have children due to being an Arithmetic expression, thus we check for children.
+	return ExpressionToColumnDefaultValue(ctx, parsedExpr, len(parsedExpr.Children()) == 0 && !strings.HasPrefix(exprStr, "("))
+}
+
+// ExpressionToColumnDefaultValue takes in an Expression and returns the equivalent ColumnDefaultValue if the expression
+// is valid for a default value. If the expression represents a literal (and not an expression that returns a literal, so "5"
+// rather than "(5)"), then the parameter "isLiteral" should be true.
+func ExpressionToColumnDefaultValue(ctx *Context, expr Expression, isLiteral bool) (*ColumnDefaultValue, error) {
+	expr, err := ColumnDefaultManager.ExpressionTransformUp(expr)
+	if err != nil {
+		return nil, err
+	}
+	//TODO: currently (2+2)/2 will, when output as a string, give (2 + 2 / 2), which is clearly wrong
+	return NewColumnDefaultValue(expr, isLiteral), nil
+}
+
+// MustStringToColumnDefaultValue is StringToColumnDefaultValue except that it panics on errors.
+func MustStringToColumnDefaultValue(ctx *Context, exprStr string) *ColumnDefaultValue {
+	expr, err := StringToColumnDefaultValue(ctx, exprStr)
+	if err != nil {
+		panic(err)
+	}
+	return expr
+}
+
+// ColumnDefaultValue is an expression representing the default value of a column. May represent both a default literal
+// and a default expression. A nil pointer is a valid instance of this struct, and all method calls will return without error.
+type ColumnDefaultValue struct {
+	Expression
+	literal bool
```

Comment:
Not sure this type is pulling its weight. Why is the distinction of being a literal important or not enough to break out a new type? Why isn't a columns' default a sql.Expression, which can be nil? What value does this type add?

And if it is a new type, it's a new type of expression and should live in the expressions package.

Reply:
I had this initially in the `expressions` package, before getting an import cycle and moved it here. The position of everything was because I ran into import cycles. The `expressions` package references the `sql` package (which this lives in), so I would be unable to directly reference this type from within `sql`.

So the literal distinction does have importance to it. `DEFAULT 2` and `DEFAULT (2)` have a different set of rules that they abide by, and that includes functions such as `NOW()` and `(NOW())` (the former is viewed under the _literal_ rules). In a nutshell, all valid expressions can be applied to all types, however literals can only apply to specific types (which depends on the literal). For example, `col INT UNSIGNED DEFAULT -1` is invalid, however `col INT UNSIGNED DEFAULT (-1)` is valid, even though it will cause an error whenever the column is not specified.

A different example would be `col VARCHAR(20) DEFAULT NOW()` vs `col VARCHAR(20) DEFAULT (NOW())`. The former will error, however the latter will not only succeed at creation, but correctly convert the date into a string on insertion.

Additionally, the integrators would only be storing a string (we could make it a string and boolean), so we want that string to parse back into the same thing (a literal or expression). When stringifying this, we can include an outer set of parentheses if it's not a literal.

The alternative to this implementation would be an expression and a boolean, but I feel they should be in the same structure since they're tied together. I'm definitely treating this more as a struct that implements `sql.Expression` rather than an expression in the `expressions` package, since this is specific to column defaults.

Reply:
I didn't understand the distinction between an expression and a literal in this context. That makes sense.

Still belongs in the expression package though. Seems like the main drawback to putting it there is that you wouldn't be able to declare it as an *expression.ColumnDefault in the sql package (because of circular imports), which is not ideal, but not a deal breaker. What do we lose by the sql package considering the field to be a generic Expression rather than this specific one?

Reply:
So, at least from an integrator's standpoint, I'm not wanting default values to be something that are only available when used on the SQL path. An integrator should be able to have their own write path, and yet also able to make use of the default outside of that. With the default value being it's own discrete type, there will be additionally methods to the type to allow its usage outside of just being a standard expression. This is all toward usability outside of the SQL engine.

At the moment, the only additional information that's stored within is whether the value is a literal or expression. I'm planning on extending that, so just an expression and boolean will not cover those additional methods. That's the reasoning for making it a discrete type. And with it being a discrete type, it cannot live in the expression package due to import cycles.